src/beamlime/applications/_nexus_helpers.py

```diff
@@ -1,155 +1,378 @@
         return structure
     head, *tail = path
-    for child in structure['children']:
+    for child in structure["children"]:
         if head == _node_name(child):
             return find_nexus_structure(child, tail)
+    raise KeyError(f"Path {path} not found in the nexus structure.")
+
 
+def find_parent(
+    structure: NexusDictStructure, path: Sequence[Optional[str]]
+) -> NexusGroupDict:
+    # TODO: We can use typeguard here later.
+    return find_nexus_structure(structure, path[:-1])
 
-def path_to_message_group(structure: Mapping, message: Mapping) -> str:
-    '''Finds the group where the message should be merged'''
-    kind, content = message
+
+def find_ev44_matching_paths(
+    structure: NexusDictStructure, data_piece: DeserializedMessage
+) -> Iterable[NexusPath]:
+    source_name = data_piece["source_name"]
     for path, node in iter_nexus_structure(structure):
         if (
-            node.get('module') == kind
-            and node.get('config', {}).get('source') == content['source_name']
+            node.get("module") == "ev44"
+            and node.get("config", {}).get("source") == source_name
         ):
-            return path
+            yield path
+
 
+def _merge_message_into_nexus_group_store(
+    *,
+    structure: NexusDictStructure,
+    nexus_group_store: NexusGroupDictStore,
+    data_piece: DeserializedMessage,
+    path_matching_func: Callable[
+        [NexusDictStructure, DeserializedMessage], Iterable[NexusPath]
+    ],
+    data_field_initialize_func: Callable[[NexusGroupDict, DeserializedMessage], None],
+    merge_func: Callable[[NexusGroupDict, DeserializedMessage], None],
+) -> None:
+    """Bridge function to merge a message into the store.
 
-def merge_message_into_store(
-    store: Mapping[Tuple[Optional[str], ...], Mapping],
-    structure: Mapping,
-    message: Tuple[str, Mapping],
+    This bridge was needed since different modules
+    have different ways to match the paths and merge the data.
+
+    Parameters
+    ----------
+    nexus_group_store:
+        The data module store that holds the data.
+    structure:
+        The nexus structure.
+    data_piece:
+        The content of the message.
+    path_matching_func:
+        A function that returns the paths that match the message.
+    data_field_initialize_func:
+        A function that initializes the datasets in the module.
+        *Initialize is done only when the relevant message arrives.*
+        It is because we should distinguish between empty dataset and
+        unexpectedly-not-receiving data.
+        It also depends on the ``data_piece`` to handle optional fields.
+    merge_func:
+        A function that merges the message into the store.
+
+    Side Effects
+    ------------
+    The ``nexus_group_store`` is updated with the merged data.
+
+    """
+    for path in path_matching_func(structure, data_piece):
+        if path not in nexus_group_store:
+            parent = find_parent(structure, path)
```

Comment:
(minor) Why make `find_parent` a separate function if it is only used here? In my opinion that just creates unnecessary indirection. The variable name already mentions that the value is the 'parent'.

Reply:
It was just for the types... I'm trying to pass the mypy as much as possible...

Reply:
Why don't mypy let it pass without the function?